presto-spi/src/main/java/com/facebook/presto/spi/type/Varchars.java

```diff
@@ -52,17 +53,44 @@ public static Slice truncateToLength(Slice slice, int maxLength)
         if (maxLength == 0) {
             return Slices.EMPTY_SLICE;
         }
+
+        int length = getTruncatedLength(slice, 0, slice.length(), maxLength);
+        if (length == slice.length()) {
+            return slice;
+        }
+        return slice.slice(0, length);
+    }
+
+    public static int getTruncatedLength(Slice slice, int offset, int length, int codePointsCount)
+    {
+        requireNonNull(slice, "slice is null");
+        if (length < 0) {
+            throw new IllegalArgumentException("length must be greater than or equal to zero");
+        }
+        if (offset < 0 || offset + length > slice.length()) {
+            throw new IllegalArgumentException("invalid offset/length");
+        }
+        if (codePointsCount < 0) {
+            throw new IllegalArgumentException("codePointsCount must be greater than or equal to zero");
+        }
+        if (codePointsCount == 0) {
+            return 0;
+        }
+
         // min codepoint size is 1 byte.
         // if size in bytes is less than the max length
         // we don't need to decode codepoints
-        int sizeInBytes = slice.length();
-        if (sizeInBytes <= maxLength) {
-            return slice;
+        if (slice.length() <= offset + codePointsCount) {
+            return length;
         }
-        int indexEnd = offsetOfCodePoint(slice, maxLength);
-        if (indexEnd < 0) {
-            return slice;
+
+        int endIndex = offsetOfCodePoint(slice, offset, codePointsCount);
```

Comment:
Yes there is a test to cover this `assertEquals(getTruncatedLength(Slices.utf8Slice("abc"), 0, 3, 5), 3);`